rclcpp/doc/notes_on_statically_typed_parameters.md

```diff
@@ -85,6 +85,16 @@ if (mode == "modeB") {
 }
 ```
 
+Or, wrap the declaration in a try-catch:
+
+```cpp
+try {
+  node->declare_parameter<int64_t>("param_that_is_optional");
+} catch (const rclcpp::exceptions::NoParameterOverrideProvided &) {
+  // a parameter value was not provided
+}
+```
```

Comment:
I think to be equivalent to the above, wouldn't it need to be like this:

```suggestion
Or, wrap the declaration in a try-catch:

```cpp
auto mode = node->declare_parameter("mode", "modeA");
try {
  node->declare_parameter<int64_t>("param_that_is_optional");
} catch (const rclcpp::exceptions::NoParameterOverrideProvided &) {
  // a parameter value was not provided
  if (mode == "modeB") {
    // parameter and parameter override was required
    throw;
  }
}
```
```

What you've described here is more like "conditionally declare a parameter if an override is given", which is like a subset of https://github.com/ros2/rclcpp/blob/7bfda87d32bab7e9caa843b2ededa18e427fd00c/rclcpp/include/rclcpp/node_options.hpp#L329-L343.

Also, I'm kind of concerned with the conditionally declared parameters. I'm not saying they shouldn't be allowed, but if I were making a system, I'd prefer the parameters to be the same no matter what, and if one was not provided (and that was ok) it should be "not set" or something. I'm thinking about "dumping" parameters from a node in various scenarios. It's maybe undesirable to have different number and order of parameters in the dump depending on the circumstances, because then if a parameter is missing from the dump, it's hard to know if it was specified or if the dump maybe came from a difference version of the program that didn't have the parameter yet/any longer. But that's just a architectural decision for the developer of the node.

Reply:
Maybe what we need is a representation for a *value* that is "not set"; we only allow the *type* to be "not set".

Reply:
Yeah, I think this is essentially what @ivanpauno was talking about with "uninitialized parameters". It can have a type, say `int64_t`, but the value is either an `int64_t` or `uninitialized`.

Reply:
I like @wjwwood suggestion [here](https://github.com/ros2/rclcpp/pull/1588#discussion_r597995432).
If not, the try/catch block isn't completely equivalent.

> Also, I'm kind of concerned with the conditionally declared parameters. I'm not saying they shouldn't be allowed, but if I were making a system, I'd prefer the parameters to be the same no matter what, and if one was not provided (and that was ok) it should be "not set" or something. I'm thinking about "dumping" parameters from a node in various scenarios.

That sounds fair.